### PR TITLE
apiserver: get rid of ReadWritePort in config

### PR DESCRIPTION
**Pull request by sttts**
Executing removal TODO by making the read write port logic explicit, and not hidden deep in the secure serving code.

Preparation for https://github.com/kubernetes/kubernetes/pull/65832

### CODE REVIEW

File: pkg/master/controller.go [hunk 1]
```diff
@@ -79,6 +79,15 @@ type Controller struct {
 
 // NewBootstrapController returns a controller for watching the core capabilities of the master
 func (c *completedConfig) NewBootstrapController(legacyRESTStorage corerest.LegacyRESTStorage, serviceClient coreclient.ServicesGetter, nsClient coreclient.NamespacesGetter, eventClient coreclient.EventsGetter) *Controller {
+	publicServicePort := 443 // this is the legacy behaviour. Without a secure port, we cannot serve the kubernetes service correctly though.
```
Comment (deads2k) [inline pkg/master/controller.go]: This seems weird.  If we require it, why not simply fail early if we have no secure port?

Comment (sttts) [reply pkg/master/controller.go]: just replicating the old behaviour here. Am fine with failing as well. In other words: do not allow zero with `--secure-port`.

Comment (hzxuzhonghu) [reply pkg/master/controller.go]: What's the behavior previously when `--secure-port` is set to 0?

Comment (sttts) [reply pkg/master/controller.go]: Then https://github.com/kubernetes/kubernetes/pull/65830/files#diff-01f8ed603e3f5cc358d0394c0ab0cb00L51 was skipped and the default value of 443 was staying.

Comment (sttts) [reply pkg/master/controller.go]: Fixed

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -359,9 +355,14 @@ func (c *Config) Complete(informers informers.SharedInformerFactory) CompletedCo
...
-	// if there is no port, and we have a ReadWritePort, use that
-	if _, _, err := net.SplitHostPort(host); err != nil && c.ReadWritePort != 0 {
-		host = net.JoinHostPort(host, strconv.Itoa(c.ReadWritePort))
+	// if there is no port, and we listen on one securely, use that one
+	if _, _, err := net.SplitHostPort(host); err != nil {
+		if _, port, err := c.SecureServing.HostPort(); err == nil {
+			host = net.JoinHostPort(host, strconv.Itoa(port))
+		} else {
+			// last way out: 443. This is the legacy behaviour for this case.
+			host = net.JoinHostPort(host, strconv.Itoa(443))
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: Can we error here in instead?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -615,3 +616,15 @@ func NewRequestInfoResolver(c *Config) *apirequest.RequestInfoFactory {
 		GrouplessAPIPrefixes: legacyAPIPrefixes,
 	}
 }
+
+func (s *SecureServingInfo) HostPort() (string, int, error) {
+	if s == nil || s.Listener == nil {
+		return "", 0, fmt.Errorf("no listener found")
+	}
+	host, portStr, err := net.SplitHostPort(s.Listener.Addr().String())
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: lost this error.

**Comment by deads2k** [comment]
minor comments, lgtm otherwise, but the lost error needs handling.

**Pull request merged**